### DISCUSSION TITLE
Backup will back up the backup directory(thus causing a doubling of size everytime)

### DISCUSSION THREAD

**Discussion by galatea99** [pocketbase/pocketbase#2548] [Q&A] [answered]
Any idea why the backup grows to almost double every time it backsup?
![Screenshot 2023-05-23 213056](https://github.com/pocketbase/pocketbase/assets/29354345/e4f8063c-35c3-4ab2-afe2-2a9231548705)

Edited: Found out reason: it's backing up the backup directory too!

**Comment by ganigeorgiev** [comment]
> Found out reason: it's backing up the backup directory too!

No, this shouldn't be the case because the backup directory is explicitly excluded from the zip (_we also have [test that ensures this](https://github.com/pocketbase/pocketbase/blob/master/core/base_backup_test.go#L110)_). You can try to download one of the above zip and see if there is a `backups` directory inside it.

I'm not sure what could be the reason for the double size increase. I guess it could be the logs?

**Reply by ganigeorgiev** [reply]
I've just updated the test to make sure that there is at least 1 backup file - https://github.com/pocketbase/pocketbase/commit/db20e38cda8709e988afeb3e4c8e21cf8bef6842 and it is passing successfully.

I've also extracted one of the generated zips locally and it contains only the db file(s) and the uploaded files `storage` dir.

Please try to extract one of your backups and inspect its content.

**Comment by galatea99** [comment]
It's strange only the first backup zip doesn't have the backup dir but the rest all got the backup folder.

**Reply by ganigeorgiev** [reply]
I'm not sure how to reproduce it. On what OS you are testing - Windows?
Could you provide a little more details on your setup (eg. is PocketBase running in a container, do you have custom pb_data with the `--dir` flag, etc.)?

**Reply by ganigeorgiev** [reply]
> only the first backup zip doesn't have the backup dir

If it doesn't have the `backups` dir why the size increase is similar to the others? Or do you mean the earliest generated backup 327MB?

**Reply by galatea99** [reply]
I tested it on windows 10.
Run it baremetal, no container. 
No custom dir, all default setting. 
I was testing the backup function locally and then will update the linux cloud version.
I have not update the linux one yet to 0.16 yet.
yes, the manual backup won't backup the "backup" dir.
Only the autobackup will backup including the "backup" dir. It's reproducible.

**Comment by galatea99** [comment]
![Screenshot 2023-05-23 220245](https://github.com/pocketbase/pocketbase/assets/29354345/dc6854c1-e340-43ec-bccc-665dad1e77b3)
I have tried it again, deleting ALL the backups.
Only the first backup will have no "backup" folder. Starting the second time it will auto backup the "backup".
Hope this helps.

**Reply by ganigeorgiev** [reply, marked as answer]
This is very frustrating. I'm not able to reproduce it. Here are couple screenshots from my environment:

![backups_list](https://i.imgur.com/bEkjmuZ.png)

![backups_zip](https://i.imgur.com/jDZpPLL.png)

I guess it could be something Windows related but I currently don't have a Windows machine to test it. Will try sometime later to install a Windows vm and we'll see if I can reproduce it.

**Comment by galatea99** [comment]
It's doing a Backup(n+1) = Backup(n) + backup(n-1).
Like a recursion!
I trust it will work on other OS, but it just happens I test it with Windows first, then push to linux.

**Reply by ganigeorgiev** [reply]
There is no such builtin business logic. Something else is going on here but I have no clue what could cause it to behave this way.

**Comment by galatea99** [comment]
![image](https://github.com/pocketbase/pocketbase/assets/29354345/6e8911bf-401f-48be-91e3-ae97bf74599d)

Yes, this is weird. OS related I guess. Can you specifically just back up 6 files instead?

**Reply by ganigeorgiev** [reply]
> Can you specifically just back up 6 files instead?

I'm confused, could you elaborate?

**Reply by galatea99** [reply]
3 data.* and 3 logs.*.
Then it won't give windows to include the backup folder accidentally.

**Reply by ganigeorgiev** [reply]
That's not a solution. We need also the local uploaded files (in case the local file system is used) + any other user defined auxiliary file/directory inside pb_data (eg. `.gitignore`).

**Comment by galatea99** [comment]
I just tried the Mac version. No problem whatsoever.
So it's the Windows version only, may have to filter out the backup dir specifically to get around.

**Comment by ganigeorgiev** [comment]
I found the culprit.

I've incorrectly assumed that the path argument in [`fs.WalkDirFunc`](https://pkg.go.dev/io/fs#WalkDirFunc) will use `os.PathSeparator` as path delimiter while it seems that it is always "/", which is causing [our exclude zip path check](https://github.com/pocketbase/pocketbase/blob/master/tools/archive/create.go#L51-L57) to fail because we always normalize the path with `filepath.Clean` (it will replace the path separators with the one for the specific OS).

I'll submit a fix sometime later this week.
